content/posts/api-contracts-definitons.md

```diff
@@ -0,0 +1,126 @@
+
+HTTP works great as a means of communication for microservices because it is open, reliable, programming language-agnostic and works great over the wire. All of these features are crucial to modern services, as they allow engineers to change the underlying technologies (e.g., change the back end code from Python to Go) without it affecting the contract. Therefore, the API's consumers don't even need to know about the implementing technology and the providing team can take independent decisions respectively.
+
+Service contracts usually contain the following four components:
+
+- Available endpoints and operations on each endpoint
+- Operation parameters Input and output for each operation
+- Authentication methods
+- Contact information, license, terms of use and other information
+
+## Manual Specification
+
+The easiest way of creating a contract is to manually write it, and then writing the respective code that should implement the contract. This is quite tedious and error-prone, as you have to basically write everything twice. When you change your implementation, you have to think about also changing the documentation and contract in the exact same way and vice versa. A way better approach is to either pick a technology that is contract-based and incorporates the interface specification in the exposed API, or to at least automate either the generation of the contract from the implementation or the other way around.
+
+## Contract First vs. Implementation First
+
+There are two basic approaches to keeping the contract and the implementation in sync in an automated way. The first one is to write the code first and have the contract generated from that. The second approach is to write the contract and have the respective implementation code generated from that.
+
+Both are viable approaches, but in general, it is preferred to write the contract first and generate implementation code from it. The reason being that when you begin implementing your service, ideally, the contract has already been defined and communicated with potential consumers of your API to allow them to work independently of your implementation. Having a human- and machine-readable contract checked into your source code repository allows you to track changes to that contract over time and additionally serves as documentation for what the implementation code does (or at least what it should do).
+
+## Technologies
```

Comment:
Contract First Technologies

Reply:
The first two technologies are not necessarily "Contract First" but can be used either manually or with a contract first or implementation first approach. I would leave it as is.